Start Conway Imp tests with an initial committee and constitution

teodanciu opened the pull request:
# Description

This PR is setting an initial committee and constitution for Conway (emulating the ones that  will be defined in genesis). 
For this, it was necessary to slightly modify the initialization of the NewEpochState in the Imp framework, to allow initialization actions before returning the state. 

Tests were adjusted to only elect a committee if this action is actually meaningful to the test (relying on the initial committee otherwise). 
Going through many tests was an opportunity to make some small  improvements: 
 * an attempt to be more disciplined with setting pparams. I have consolidated a few of them 
 * updating some functions to reduce the noise of converting between KeyHash and Credential
 * removing the duplicated inclusion of a test 
 
<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

teodanciu committed:
Adjust conway Imp tests to an existent initial committee

teodanciu committed:
Remove duplicated inclusion of `GovCert` spec in conway Imp tests

teodanciu committed:
Update `setupSingleDRep` to return Credential instead of KeyHash

in order to reduce noise in tests

teodanciu committed:
Consolidate setting of pparams in Imp tests

* set sensible values when initializing NewEpochState
* remove duplicated modifications of the same parameters

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -212,6 +214,18 @@ conwayModifyPParams f = modifyNES $ \nes ->
+initialCommitteeKeyPair = mkKeyPair 30
+
+initialCommitteeCred :: forall c. Crypto c => Credential 'ColdCommitteeRole c
+initialCommitteeCred = KeyHashObj (hashKey @c (vKey initialCommitteeKeyPair))
```
Global bindings like that are not a good idea.

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -222,16 +236,21 @@ instance
           def
             & rsEnactStateL .~ mkEnactState (epochState ^. epochStateGovStateL)
-     in nes & nesEsL .~ setCompleteDRepPulsingState def ratifyState epochState
+    void $ addKeyPair initialCommitteeKeyPair
```
Let's generate it at random. You'll just need to add `MonadGen`  constraint to `initImpNES`
```suggestion
    committeeKeyPair <- freshKeyPair
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -268,6 +287,19 @@ instance
+  ConwayEraImp era =>
+  ImpTestM era (NonEmpty (Credential 'HotCommitteeRole (EraCrypto era)))
+registerInitialCommittee = do
+  hk <- registerCommitteeHotKey initialCommitteeCred
```
Global bindings like that is a bad idea. Look up the committee cold credential in the state instead

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -268,6 +287,19 @@ instance
+getInitialCommitteMembers ::
+  Era era =>
+  ImpTestM era (NonEmpty (Credential 'ColdCommitteeRole (EraCrypto era)))
+getInitialCommitteMembers = pure $ initialCommitteeCred :| []
```
Ditto. Look it up in the state

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -212,6 +214,18 @@ conwayModifyPParams f = modifyNES $ \nes ->
           DRComplete snapshot (ratifyState & rsEnactStateL . ensCurPParamsL %~ f)
 
+initialCommitteeKeyPair :: Crypto c => KeyPair 'ColdCommitteeRole c
+initialCommitteeKeyPair = mkKeyPair 30
```
These can be generated randomly, no need to use this hack of relying on an integer.

See my comment below on how to do that

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -212,6 +214,18 @@ conwayModifyPParams f = modifyNES $ \nes ->
+initialCommitteeCred = KeyHashObj (hashKey @c (vKey initialCommitteeKeyPair))
+
+initialCommittee :: forall era. Era era => Committee era
+initialCommittee =
```
Ditto. Global binding that needs to be removed

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -997,21 +1029,8 @@ electBasicCommittee ::
-      & ppDRepVotingThresholdsL
-        %~ ( \dvt ->
-              dvt
-                { dvtCommitteeNormal = 1 %! 1
-                , dvtCommitteeNoConfidence = 1 %! 2
-                , dvtUpdateToConstitution = 1 %! 2
-                }
```
Looks like we lost all of those threshold in the setup of every test that used to `electBasicCommittee`? Are they set somewhere else that I am missing?

teodanciu replied:
Indeed! I forgot that these were set here!  Will make sure they are set, thanks!

lehins commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -375,7 +376,8 @@ class
   where
-  initImpNES :: NewEpochState era
+  initImpNES ::
+    (Era era, MonadState (ImpTestState era) m) => m (NewEpochState era)
```
I was thinking more about it and this turns out to be a lot more general function. So, let's rename it and change its semantics to reflect that:
```suggestion
  initImpTestState :: (MonadState (ImpTestState era) m, MonadGen m) => m ()
```

Important notes:

* `Era` is redundant
* `MonadGen` is useful, as mentioned in my other comment
* This action should not return anything, but instead should modify the state with all the initialization that needs to be done.

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -222,16 +236,21 @@ instance
             & nesEsL . curPParamsEpochStateL . ppGovActionLifetimeL .~ EpochInterval 30
+            & nesEsL . curPParamsEpochStateL . ppCommitteeMaxTermLengthL .~ EpochInterval 20
+            & nesEsL . epochStateGovStateL . committeeGovStateL .~ SJust initialCommittee
+        -- TODO: configure sensible voting thresholds, rather than defaulting to zero
```
From the looks of it, this PR relaxes a bunch of tests by removing thresholds, which used to be set by a call to `electBasicCommitee`.

I think we need to take care of this TODO in this PR or at the very least make sure we have the same thresholds in tests as before this PR (which probably would mean resetting them to zero in some tests that relied on zero thresholds)

lehins reviewed (changes requested):
Looks good, but committee initialization is a bit hacky and it seems that thresholds got relaxed for many tests

teodanciu committed:
Change imp initialization to set the the initialized state in MonadState

teodanciu committed:
Initialize Conway NewEpochState with an initial committee

teodanciu committed:
Add initial constitution in conway ImpTestState initialization

teodanciu commented:
* Removed global bindings for committee, and generated them with `freshKeyPair` instead
* Set voting thresholds when initializing and adjusted the tests accordingly
* Added initial constitution 

If you want to have another look @lehins when you can, thanks!

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -224,16 +227,36 @@ instance
-            & rsEnactStateL .~ mkEnactState (epochState ^. epochStateGovStateL)
-     in nes & nesEsL .~ setCompleteDRepPulsingState def ratifyState epochState
+  initImpTestState =
+    do
```
We can make this change in a subsequent PR. Just pointing it out as an inconsistency :wink: 
```suggestion
  initImpTestState = do
```

teodanciu replied:
Fixed, thanks!

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -224,16 +227,36 @@ instance
+        let newNes =
+              (initAlonzoImpNES nes)
+                & nesEsL . curPParamsEpochStateL . ppDRepActivityL .~ EpochInterval 100
+                & nesEsL . curPParamsEpochStateL . ppGovActionLifetimeL .~ EpochInterval 30
```
`electBasicCommittee` used to also set the deposit:
```suggestion
                & nesEsL . curPParamsEpochStateL . ppGovActionLifetimeL .~ EpochInterval 30
                & nesEsL . curPParamsEpochStateL . ppGovActionDepositL .~ Coin 123
```

lehins reviewed (approved):
Looks great. Thank you!
